### PR TITLE
Support for opt-in network isolation in build/test sandboxes

### PR DESCRIPTION
**Pull request by alebcay**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [x] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [x] Have you successfully run `brew typecheck` with your changes locally?
- [x] Have you successfully run `brew tests` with your changes locally?

-----

This is probably going to need a lot of iterating, but I think it's in a state that's ready for some eyes and feedback now.

## Overview

This PR introduces opt-in (disabled by default) support for network isolation during build, test, and postinstall. A formula can opt-in like so:

```ruby
class Hello < Formula
  url "[...]"
  sha256 "[...]"
  [...]
  # For enabling on a single phase:
  enable_offline_phase :build
  # For enabling on multiple phases:
  enable_offline_phase [:build, :postinstall]
  # Without arguments, enables on all phases:
  enable_offline_phase

  bottle do
    [...]
  end
  [...]
```

Suppose we opt-in a formula for offline build phase like so:

```ruby
class Hello < Formula
  desc "Program providing model for GNU coding standards and practices"
  homepage "https://www.gnu.org/software/hello/"
  url "https://ftp.gnu.org/gnu/hello/hello-2.12.1.tar.gz"
  sha256 "8d99142afd92576f30b0cd7cb42a8dc6809998bc5d607d88761f512e26c7db20"
  license "GPL-3.0-or-later"
  enable_offline_phase :build
  [...]
```

and perhaps there's a pesky `curl` call in the install block:

```ruby
def install
    system "curl", "example.org"
    ENV.append "LDFLAGS", "-liconv" if OS.mac?
[...]
```

When we attempt to install this formula:

```
==> curl example.org
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0curl: (6) Could not resolve host: example.org
[...]
Error: hello 2.12.1 did not build
Logs:
     /Users/caleb/Library/Logs/Homebrew/hello/00.options.out
     /Users/caleb/Library/Logs/Homebrew/hello/01.curl
READ THIS: https://docs.brew.sh/Troubleshooting
```

## Things still to be (re-)considered

- I refer to "phases" here since it was the first term that came to mind. Any opinions on what to call "build", "test", "postinstall", etc.? This naming would be reflected in the method naming.
  - Open generally to feedback on the method/DSL naming.
- Going in a different direction, is the option to choose phases for which to enable this too granular? Maybe we just settle for one big on/off toggle. But then there's a lot of network-related tools/libraries that probably don't need connectivity to build but might need it for a meaningful test.
  - Compromise: combine `:build` and `:postinstall` into `:build`, while leaving `:test` as its own
  - Maybe we can open some exceptions for localhost in the sandbox, at least when running test block
- Running `enable_offline_phase` without arguments currently enables for all. Would it be less surprising for the no-arguments case to enable for build only? build + postinstall? Or maybe require an argument
  - Consider providing `:all` as an easy way to enable for everything, if we decide to require an argument

### CODE REVIEW

**Comment by RandomDSdevel** [comment]
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;What about having:  

 1. An '`offline do`' block for use when only part(s) of an '`install`' (build,) '`postinstall`,' or '`test do`' block need network isolation; 
 1. An '`:offline`' '`Symbol`' argument that formula authors can pass to `Formula`'s '`install`,' '`postinstall`,' and '`test`' methods before their block argument that causes the entire block to get evaluated as if all of its author-sipplied contents were written inside of such an '`offline do`' block; and 
 1. some kind of DSL to specify that all of a formula's '`install`,' '`postinstall`,' and '`test`' methods are to be evaluated as if they were all passed such an '`offline`' argument?  

    &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;This could be by either:  

    - Calling '`Formula`' method, like the '`enable_offline_phase!`' method that this PR introduces in its current form, or 
    - Inheritance from a new '`Formula`' subclass (maybe call it `OfflineFormula`.'

**Comment by alebcay** [comment]
> An '`offline do`' block for use when only part(s) of an '`install`' (build,) '`postinstall`,' or '`test do`' block need network isolation;

For this point in particular, to achieve this would require more significant re-design of the formula installation/postinstall/test sandbox setup code. Currently, the sandbox is set up before the respective block is run, so the decision/logic to enable this has to be outside the install/test/postinstall blocks; trying to change it inside the block is too late since everything inside e.g. `def install` is already inside the sandbox.

> An ':offline' 'Symbol' argument that formula authors can pass to Formula's 'install,' 'postinstall,' and 'test' methods before their block argument that causes the entire block to get evaluated as if all of its author-supplied contents were written inside of such an 'offline do' block

I personally like this idea but am not sure how it could be achieved syntactically.

**Comment by RandomDSdevel** [comment]
>> An '`offline do`' block for use when only part(s) of an '`install`' (build,) '`postinstall`,' or '`test do`' block need network isolation;
> 
> For this point in particular, to achieve this would require more significant re-design of the formula installation/postinstall/test sandbox setup code. Currently, the sandbox is set up before the respective block is run, so the decision/logic to enable this has to be outside the install/test/postinstall blocks; trying to change it inside the block is too late since everything inside e.g. `def install` is already inside the sandbox.
> 
> ⋮

&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Ah, right; definitely fair enough.  Never mind that idea, then, at least for the foreseeable future.  

> ⋮
> 
>> An '`:offline`' '`Symbol`' argument that formula authors can pass to `Formula`'s '‘install`,' '`postinstall`,' and '`test`' methods before their block argument that causes the entire block to get evaluated as if all of its author-supplied contents were written inside of such an '`offline do`' block
> 
> I personally like this idea but am not sure how it (`def install :offline`) could be achieved syntactically.

&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Maybe look at how '`it "", :needs_network do`' and '`RSpec.describe klass_or_string, :needs_network do`' are implemented in/for '`Library/Homebrew/test/`,' though I don't know if that helps since those are part of the test suite helpers.  I think they're in:  

https://github.com/Homebrew/brew/blob/master/Library/Homebrew/test/spec_helper.rb#L116-L126

https://github.com/Homebrew/brew/blob/master/Library/Homebrew/test/spec_helper.rb#L158-L169

That's what gave me that idea.

File: Library/Homebrew/dev-cmd/test.rb
Comment (MikeMcQuaid) [inline Library/Homebrew/dev-cmd/test.rb:97]: Feels like it might make sense to combine these calls given they are in the same order and the same every time?

**Comment by MikeMcQuaid** [comment]
> This PR introduces opt-in (disabled by default) support for network isolation during build, test, and postinstall. A formula can opt-in like so:

@alebcay AMAZING work here 👏🏻. Seen many attempts over the years and this seems by far the most likely to actually get adopted.

> I refer to "phases" here since it was the first term that came to mind. Any opinions on what to call "build", "test", "postinstall", etc.? This naming would be reflected in the method naming.
> * Open generally to feedback on the method/DSL naming.

Given there's only three: I wonder about just having them all be separate calls? I guess "methods" would work instead of "phases". I don't think we really describe these to users/contributors anywhere 🤔 

> * But then there's a lot of network-related tools/libraries that probably don't need connectivity to build but might need it for a meaningful test.

Yeh, I think different phases makes sense.

> * Maybe we can open some exceptions for localhost in the sandbox, at least when running test block

This makes sense to me 👍🏻 

---

The main question I have here isn't actually about the implementation but about formulae: how many formulae have you tried this on already? I'd be interested to see at least 3+ where it's working for (each of) build, postinstall, test before this gets merged.

Also I think it'd make sense pretty quickly to have some sort of environment variable to enable this without the DSL. It'd be great to be able to play around locally or with test-bot to find out which formulae can have this DSL used. We also probably want to e.g. require it unconditionally for new formulae at some point.

Relatedly, based on similar DSL addition work like this, it might be nice to have effectively a no-op reverse DSL e.g. `disable_offline_phase`. This might seem odd but: if this merges, eventually we'll end up flipping the default on formulae to disabling network access unless specifically enabled (and maybe in ~5-10 years: we remove the ability to disable here entirely).

This DSL would also make it easier to have a triage process (perhaps with the environment variable above) for formulae where:
- neither DSL used: formula hasn't been checked yet
- `enable_*` used: formula has been checked and offline works (for at least one method)
- `disable_*` used: formula has been checked and offline is broken (for at least one method)

Again: great work here and looking forward to seeing how this evolves! ❤️

**Comment by MikeMcQuaid** [comment]
@RandomDSdevel I appreciate the enthusiasm but I'm not sure your code review on PRs like this is helpful given your knowledge of Homebrew internals and Ruby, sorry ❤️

File: Library/Homebrew/formula.rb
Comment (carlocab) [inline Library/Homebrew/formula.rb:3124]: Maybe call it `deny_network_access`? It's a bit more obvious what it does, and implies a natural name for the opposite (`allow_network_access`) for when we start making it the default.

**Review by carlocab — APPROVED** [review]
Minor bikeshedding comment, but otherwise seems reasonable to me

File: Library/Homebrew/formula.rb
Comment (woodruffw) [reply Library/Homebrew/formula.rb:3124]: +1 on `deny_network_access` -- I think that also makes the default case ("deny all with nothing explcitly listed") a little more intelligible 🙂

**Comment by alebcay** [comment]
Thanks for the feedback everyone, made some adjustments:
- DSL is now `{allow,deny}_network_access_for_{build,postinstall,test}` - total 6 methods
- Added environment variables `HOMEBREW_FORMULA_{BUILD,POSTINSTALL,TEST}_NETWORK`. Set to `allow` or `deny` to force allow or deny, respectively; otherwise if unset will default to the value(s) specified via DSL in the formula(e). Open to further bikeshedding on the name of these env vars.

One thing I noticed - **tests that use resources currently cannot run offline because, unlike formula build time, the resource fetch seems to be executed in the sandbox (vs before the sandbox)**.

Todo - consider opening up localhost in test sandbox when network access is denied.

---

Some formulae that already seem to be working offline - or not - with this. Might update with some more in the next day or two.

| Formula | Build succeeds? | Postinstall succeeds? | Test succeeds? |
| --- | --- | --- | --- |
| `cmake` | ✅ | ✅ | ✅ |
| `ca-certificates` | ✅ | ✅ | ✅ |
| `xz` | ✅ | ✅ | ✅ |
| `gh` | ❌<sup>1</sup> | ✅ | ✅ |

<sup>1</sup> - `go build` needs network access to pull modules/dependencies

File: Library/Homebrew/formula.rb
Comment (MikeMcQuaid) [reply Library/Homebrew/formula.rb]: Yeh, I like that too.

@alebcay I know this seems bizarre but there may be a case (e.g. in `brew tests`) to deny network access during `fetch`!

**Comment by MikeMcQuaid** [comment]
> In the long term, I think there will be formulae (like `gh` above) that will never be able to build offline with our current processes/policies - we would have to do a lot of vendoring to control downloading stuff ourselves (kind of like we do for some Python formulae).

I don't think we should ever do that vendoring. I think better to make some DSL changes/additions. Most of these dependency download systems use checksummed and/or version pinned downloads here and have their own (used at comparable scale to Homebrew) security around this. Instead we should isolate what they do so they aren't able to e.g. arbitrarily write and perhaps restrict which hosts they can download from.

I think there's probably a place for a `preinstall do` or `fetch do` and/or `pretest do` that has different sandboxing:
- it can download arbitrary stuff off the internet
- it can _only_ write to the sandboxed temporary directory
- maybe even: it can _only_ read from the sandboxed temporary directory
- this temporary directory is then copied into the install directory to be used by an offline-only installation

To be clear: this should not block this PR being merged/DSL being added but will be worth thinking about as time goes on and we want to get 100% of homebrew-core formulae offline.

---

Alternatively, for tests: there was also talk in other issues about having resources assigned as "test resources" (much like how "normal" resources are implicitly "build resources") and then they could be fetched by `brew fetch --test` or similar. Again: not in this PR but an approach that will let us move forwards here.

File: Library/Homebrew/sandbox.rb
Comment (MikeMcQuaid) [inline Library/Homebrew/sandbox.rb:94]: Change `options = {}` into a proper Ruby attributes for each option instead. Adds better type handling with Sorbet (and, ideally, type signatures should be added as much as possible).

**Comment by RandomDSdevel** [comment]
@alebcay specifically re:  the second part of https://github.com/Homebrew/brew/pull/17081#issuecomment-2058179803, with the beginnings of a formula triage list:  

&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;I may know of some formulas that would fail to build with network isolation, but it's been a while since I've looked at them, so they may need checking to confirm that this is still the case first.  I can list them either here in this thread or in a separate Homebrew/homebrew-core triage issue thread if(/when?) one gets started.  

---

> I don't think we should ever do that vendoring.  … Most of these dependency download systems use checksummed and/or version pinned downloads here and have their own (used at comparable scale to Homebrew) security around this.

&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;I could see users potentially wanting to do this for at least some formulas in their own, third-party taps, but that's indeed outside the scope claimed by the core tap.  

> … I think better to make some DSL changes/additions.  …
> 
> ⁝
> 
> I think there's probably a place for a `preinstall do` or `fetch do` and/or `pretest do` that has different sandboxing:
> 
> * it can download arbitrary stuff off the internet
> * it can _only_ write to the sandboxed temporary directory
> * maybe even: it can _only_ read from the sandboxed temporary directory
> * this temporary directory is then copied into the install directory to be used by an offline-only installation
> 
> To be clear: this should not block this PR being merged/DSL being added but will be worth thinking about as time goes on and we want to get 100% of homebrew-core formulae offline.
> 
> ⁝

&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Build-time fetch steps confined to this new block's scope would happen after:  

 1. formula source download and resource fetch steps finish and 
 1. their results get staged.  

Maybe another potential name for the new formula stanza might be '`poststage`,' then?  

---

&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;That's the last feedback I have to add here.

**Comment by MikeMcQuaid** [comment]
> That's the last feedback I have to add here.

@RandomDSdevel I asked you earlier so I'll be a little more firm here: until you're willing to step up contributions such that you're asking to be a Homebrew maintainer: please stop commenting on PRs, thanks.

**Comment by carlocab** [comment]
> > In the long term, I think there will be formulae (like `gh` above) that will never be able to build offline with our current processes/policies - we would have to do a lot of vendoring to control downloading stuff ourselves (kind of like we do for some Python formulae).
> 
> I don't think we should ever do that vendoring. I think better to make some DSL changes/additions.

Agreed. It would probably be better to run something like `go mod vendor` (perhaps automatically?) in the source directory before sandboxing happens or something like that. (Which admittedly may be complex, since I'm not sure whether source extraction happens inside or outside the sandbox).

File: Library/Homebrew/formula.rb
Comment (carlocab) [inline Library/Homebrew/formula.rb:3151]: Not a super strong opinion, but I preferred what you did originally here, except for the name. i.e. you can call `deny_network_access` if you want to deny it everywhere, or `deny_network_access [:build, :postinstall]` if you want to allow it during `brew test`.

**Comment by alebcay** [comment]
Made some adjustments:
- DSL is now reduced back to two methods, `{allow,deny}_network_access`, accepting a single symbol or array of symbols. If no argument is passed, the method applies to all phases.
- Added typing for the new methods in `sandbox`. It's a bit messy since I tried to limit changes to only the new methods, when it would probably be less jarring once some of the existing methods are also typed. I have another branch with more of those changes, if we would prefer to have that in a PR first and then add this.

File: Library/Homebrew/env_config.rb
Comment (MikeMcQuaid) [inline Library/Homebrew/env_config.rb:233]:
```suggestion
                     'Must be "allow" or "deny".',
```

File: Library/Homebrew/env_config.rb
Comment (MikeMcQuaid) [inline Library/Homebrew/env_config.rb:232]: Unclear what the default behaviour is without this set. May also be worth mentioning if this overrides the DSL usage.

File: Library/Homebrew/formula.rb
Comment (MikeMcQuaid) [inline Library/Homebrew/formula.rb:462]: What's this look like without the disable? I think it's worth not disabling and just doing whatever that form looks like. Alternatively, if it really has to be disabled, please add a Ruby comment explaining why.

File: Library/Homebrew/formula.rb
Comment (MikeMcQuaid) [inline Library/Homebrew/formula.rb:3035]:
```suggestion
        @supported_network_access_phases = [:build, :test, :postinstall].freeze
```

or something. Also feels like this might be nicer as a constant?

File: Library/Homebrew/formula.rb
Comment (MikeMcQuaid) [inline Library/Homebrew/formula.rb:3124]: Might be worth splitting this `zip`/`length`/`to_h` given it's used elsewhere and is a little hard to understand as-is.

File: Library/Homebrew/formula.rb
Comment (MikeMcQuaid) [inline Library/Homebrew/formula.rb:3121]: A `sig` for these new methods would be nice.

File: Library/Homebrew/sandbox.rb
Comment (MikeMcQuaid) [inline Library/Homebrew/sandbox.rb:252]: A single disable and a comment explaining why would be good here.

**Review by MikeMcQuaid — COMMENTED** [review]
Looking good @alebcay, almost ready to merge here! Might be worth a homebrew/core PR to modify a few formulae so we can see what the syntax might look like there.

File: Library/Homebrew/formula.rb
Comment (alebcay) [reply Library/Homebrew/formula.rb:462]: If I run `brew --style fix` without disable:

```diff
--- a/Library/Homebrew/formula.rb
+++ b/Library/Homebrew/formula.rb
@@ -459,11 +459,11 @@ class Formula
   # @see .version
   delegate version: :active_spec

-  delegate [ # rubocop:disable Layout/HashAlignment
+  delegate [
     :allow_network_access,
     :deny_network_access,
     :network_access_allowed?,
-  ] => :"self.class"
+  ]                                                                                                 => :"self.class"

   # Whether this formula was loaded using the formulae.brew.sh API
   # @!method loaded_from_api?
   ```
   
I suspect this is why the check is also disabled already for an earlier spot (line 403). Not sure how to describe this other than "Stop RuboCop from erroneously indenting hash target" or something like that.

File: Library/Homebrew/formula.rb
Comment (alebcay) [reply Library/Homebrew/formula.rb:3129]: I had a think about how to `sig` these; unfortunately, if we add typing on this then we cannot accept plain symbols as arguments. This is because [Sorbet doesn't support literal types](https://github.com/sorbet/sorbet/issues/2504) (related - https://blog.jez.io/tenum-pro-con/).

Instead the argument would be an instance of a `T::Enum` and calling it might look something like `deny_network_access NetworkAccessPhase.Build` or `allow_network_access [NetworkAccessPhase.Test, NetworkAccessPhase.Postinstall]`.

Comment (MikeMcQuaid) [reply Library/Homebrew/formula.rb:3129]: @alebcay Could this not be `phases: T.any(T::Array[Symbol], Symbol)`?

> calling it might look something like `deny_network_access NetworkAccessPhase.Build` or `allow_network_access [NetworkAccessPhase.Test, NetworkAccessPhase.Postinstall]`.

Would rather avoid this in formulae.

Comment (alebcay) [reply Library/Homebrew/formula.rb:3129]: > @alebcay Could this not be `phases: T.any(T::Array[Symbol], Symbol)`?
> 

Strangely this is working here... not sure why something similar didn't work when I was trying to add types in `sandbox.rb`. Will try and re-visit that there too.

Comment (alebcay) [reply Library/Homebrew/formula.rb:3130]: I've changed the sigs in `formula.rb` and also adjusted some sigs in `sandbox.rb` to remove the need for `T::Enum`.

One slight downside is that Sorbet will not check if a given symbol is a valid argument or not, but that's not an assurance we had without typing, anyways.

File: Library/Homebrew/sandbox.rb
Comment (MikeMcQuaid) [inline Library/Homebrew/sandbox.rb:94]: Why do these need to be `nilable`?

**Review by MikeMcQuaid — COMMENTED** [review]
One question and then should be good to go!

**Comment by MikeMcQuaid** [comment]
@alebcay Would love to see an example PR in homebrew-core!

File: Library/Homebrew/sandbox.rb
Comment (alebcay) [reply Library/Homebrew/sandbox.rb]: You're right, these don't need to be `nilable` (copied/adapted from some temporary sig I had for `allow_write`, which _does_ need `nilable` in its current state).

**Comment by alebcay** [comment]
> @alebcay Would love to see an example PR in homebrew-core!

Opened https://github.com/Homebrew/homebrew-core/pull/169720. A few more things I thought of as I opened that:
- Do we need to enforce block/statement order for this? (If so I need to figure out how to write a cop for this). I intuitively, without any particular reason, inserted it before `def install` in this case but open to thoughts on if there's some place that we think makes more sense.
- Any considerations needed for this to work when building from source using API? Not something I have tested (and not quite sure how to test this).

**Comment by MikeMcQuaid** [comment]
> * Do we need to enforce block/statement order for this? (If so I need to figure out how to write a cop for this).

There's already an existing cop so you should be able to add it to that 👍🏻 

> * inserted it before `def install` in this case

This makes sense to me.

> * Any considerations needed for this to work when building from source using API?

Feels like it should be the same building from source/building bottles with/without API.

File: Library/Homebrew/formula.rb
Comment (MikeMcQuaid) [inline Library/Homebrew/formula.rb:471]:
```suggestion
    :allow_network_access!,
    :deny_network_access!,
```

for method names maybe (based on https://github.com/Homebrew/homebrew-core/pull/169720). happy to be overruled if you/others disagree!

Comment (woodruffw) [reply Library/Homebrew/formula.rb:471]: +1 from me on adding `!` -- IMO it makes it a bit clearer that this mutates the global "environment" the formula's stages operate in 🙂

**Review by MikeMcQuaid — APPROVED** [review]
Thanks again @alebcay, great work here!

**Pull request merged**
